### PR TITLE
Migrate to CircleCI 2.0

### PR DESCRIPTION
**Pull request by adidahiya**
Changes in this PR:

- Remove `circle.yml` in favor of `.circleci/config.yml` to configure [CircleCI 2.0](https://circleci.com/docs/2.0/)
- Optimize CI build pipeline to reduce build time from 10min -> 4min
- Move `generate-docs-data` and the resulting JSON it produces into a new package called `@blueprintjs/docs-data` so that we can have a clean separation of `typescript` versions (`documentalist > ts-quick-docs` still requires TS 2.1, while all other compilation requires TS 2.4)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  test-datetime:
+    docker:
+      - image: circleci/node:latest
+    steps:
+      - checkout
+      - attach_workspace:
+          at: '.'
+      - restore_cache:
+          key: v1-dependency-cache-{{ checksum "yarn.lock" }}
+      - run: ./node_modules/.bin/lerna run test --scope '@blueprintjs/datetime'
```
Comment (giladgray) [inline .circleci/config.yml]: `yarn lerna run test ...args`?

`yarn <binary>` is equivalent to `./node_modules/.bin/<binary>`

Comment (adidahiya) [reply .circleci/config.yml]: yeah makes sense

**Comment by giladgray** [comment]
build failures look like a typescript version mismatch? 
it's complaining about default generics syntax `Foo<T = {}>`.

**Comment by adidahiya** [comment]
getting closer...
image

**Pull request merged**